src/ripple/test/impl/JSONRPCClient.cpp

```diff
@@ -0,0 +1,164 @@
+            using boost::asio::ip::address_v4;
+            if(*pp.ip == address_v4{0x00000000})
+                *pp.ip = address_v4{0x7f000001};
+            return { *pp.ip, *pp.port };
+        }
+        throw std::runtime_error("Missing HTTP port");
+    }
+
+    template <class ConstBufferSequence>
+    static
+    std::string
+    buffer_string (ConstBufferSequence const& b)
+    {
+        using namespace boost::asio;
+        std::string s;
+        s.resize(buffer_size(b));
+        buffer_copy(buffer(&s[0], s.size()), b);
+        return s;
+    }
+
+    boost::asio::io_service ios_;
+    boost::asio::ip::tcp::socket stream_;
+    boost::asio::streambuf bin_;
+    beast::asio::streambuf bout_;
+
+public:
+    explicit
+    JSONRPCClient(Config const& cfg)
+        : stream_(ios_)
+    {
+        stream_.connect(getEndpoint(cfg));
+    }
+
+    ~JSONRPCClient() override
+    {
+        //stream_.shutdown(boost::asio::ip::tcp::socket::shutdown_both);
+        //stream_.close();
+    }
+
+    /*
+        Return value is an Object type with up to three keys:
+            status
+            error
+            result
+    */
+    Json::Value
+    invoke(std::string const& cmd,
+        Json::Value const& params) override
+    {
+        std::string s;
+        {
+            Json::Value jr;
+            jr["method"] = cmd;
+            if(params)
+            {
+                Json::Value& ja = jr["params"] = Json::arrayValue;
+                ja.append(params);
+            }
+            s = to_string(jr);
+        }
+
+        using namespace boost::asio;
+        using namespace std::string_literals;
+        std::string r;
+        r =
+            "POST / HTTP/1.1\r\n"
+            "Host: me\r\n"
+            "Connection: Keep-Alive\r\n"s +
+            "Content-Type: application/json; charset=UTF-8\r\n"s +
+            "Content-Length: " + std::to_string(s.size()) + "\r\n"
+            "\r\n";
+        write(stream_, buffer(r));
+        write(stream_, buffer(s));
+
+        read_until(stream_, bin_, "\r\n\r\n");
+        beast::asio::streambuf body;
+        beast::http::message m;
+        beast::http::parser p(
+            [&](void const* data, std::size_t size)
+            {
+                body.commit(buffer_copy(
+                    body.prepare(size), const_buffer(data, size)));
+            }, m, false);
+
+        for(;;)
+        {
+            auto const result = p.write(bin_.data());
+            if (result.first)
+                throw result.first;
+            bin_.consume(result.second);
+            if(p.complete())
+                break;
+            bin_.commit(stream_.read_some(
+                bin_.prepare(1024)));
+        }
+
+        Json::Reader jr;
+        Json::Value jv;
+        jr.parse(buffer_string(body.data()), jv);
+        if(jv["result"]["error"])
+            jv["error"] = jv["result"]["error"];
+        if(jv["result"]["status"])
+            jv["status"] = jv["result"]["status"];
```

Comment:
I didn't realize it until I tested for `"error"`'s absence, but this will create the item as null if it doesn't already exist. This breaks `RPC::contains_error`. I've fixed it in https://github.com/ximinez/rippled/commit/fbc3180578ab0a4d787912c56afbe965c0adfbd0, which is going to get merged as part of #1489, so you don't need to do anything.